### PR TITLE
BAU: Test-driven development

### PR DESCRIPTION
**Pull request by mrwilson**
Content explaining what TDD is and why we advocate for it

### CODE REVIEW

File: source/standards/test-driven-development.html.md.erb [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+Kent Beck, the author of Extreme Programming Explained, describes [TDD][canon-tdd] as .
+
+The process starts with a set of desired behaviours for a piece of code. This list could include happy-path behaviours, error cases, and interaction points with other components.
+
+The inner-loop of TDD is also commonly referred to as "Red-Green-Refactor"; red to represent a failing test, green to represent all passing tests, and refactoring only when all tests are passing.
+
+<%= partial 'partials/svg-tdd-red-green-refactor' %>
+
+### Red
+Turn exactly one item on the list into an actual, concrete, runnable test. This test should fail when you run it.
```
Comment (BeckaL) [inline source/standards/test-driven-development.html.md.erb:21]: It might not be worth complicating this, but a possible addendum to this is "it should fail for the reason you think it should fail" i.e. not compilation errors, not some unexpected reason that might mean you've made a mistake in writing the test.

Comment (mrwilson) [reply source/standards/test-driven-development.html.md.erb:21]: Good shout - I remember someone at my previous job naming this "calling your shots", and that unexpected failures were a sign of an implementation gap and also a good learning opportunity. 😂

File: source/standards/test-driven-development.html.md.erb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+### Refactor
+
+Improve the design and implementation of the code by removing duplication, utilising well known patterns and structures.
+
+
+These steps are repeated until your set of behaviours is exhausted.
+
+## Why we recommend it
+
+[canon-tdd]: https://tidyfirst.substack.com/p/canon-tdd
```
Comment (BeckaL) [inline source/standards/test-driven-development.html.md.erb:36]: Had a look at this link, it doesn't say much about the "why", mostly reiterates the process

Comment (mrwilson) [reply source/standards/test-driven-development.html.md.erb:49]: So the intention of this link was just to describe the process (as the context of the link above is "what if we can't follow the process") - it was originally further up but I forgot to push my most recent changes, which I've just done!

**Comment by james-sheppard-gov-uk** [comment]
I very much support adding TDD to The GDS Way! When I learned about TDD it was with a peer group who benefitted from a lot of time and support from our managers, which included regular, safe opportunities to practise our TDD skills with expert advice. I wonder if guidance is needed for those who are new to TDD and/or haven't had any training or coaching in it?

File: source/partials/_svg-tdd-red-green-refactor.html.erb [hunk 3]
```diff
@@ -0,0 +1 @@
..."/><path fill="#000000" stroke="#000000" stroke-width="4.0" stroke-linecap="butt" d="m349.7874 88.637115l18.152374 -6.6069336l-18.152374 -6.606926z" fill-rule="evenodd"/><path fill="#000000" fill-opacity="0.0" d="m308.38586 244.62492l85.73227 -113.44882" fill-rule="evenodd"/><path stroke="#000000" stroke-width="4.0" stroke-linejoin="round" stroke-linecap="butt" d="m322.85553 225.47736l56.79297 -75.15369" fill-rule="evenodd"/><path fill="#000000" stroke="#000000" stroke-width="4.0" stroke-linecap="butt" d="m317.5844 221.49402l-5.673004 18.465591l16.21521 -10.498932z" fill-rule="evenodd"/><path fill="#000000" stroke="#000000" stroke-width="4.0" stroke-linecap="butt" d="m384.91962 154.30699l5.673004 -18.465576l-16.21524 10.498917z" fill-rule="evenodd"/><path fill="#000000" fill-opacity="0.0" d="m210.10359 244.62492l-85.73229 -113.44882" fill-rule="evenodd"/><path stroke="#000000" stroke-width="4.0" stroke-linejoin="round" stroke-linecap="butt" d="m210.10359 244.62492l-71.262634 -94.301254" fill-rule="evenodd"/><path fill="#000000" stroke="#000000" stroke-width="4.0" stroke-linecap="butt" d="m144.11206 146.34033l-16.215225 -10.498917l5.673004 18.465576z" fill-rule="evenodd"/></g></svg>
```
Comment (huwd) [inline source/partials/_svg-tdd-red-green-refactor.html.erb:1]: if you're just putting up html, you can drop the `.erb` ending for this file. You only need that if you have ruby gubbins inside.

Also GitHub is complaining about the lack of a final line break.

File: source/standards/test-driven-development.html.md.erb [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+Test-driven development (TDD) is a core agile software development practice which aims to build flexibility and correctness into production software.
+
+The process starts with a set of desired behaviours for a piece of code. This list could include happy-path behaviours, error cases, and interaction points with other components.
+
+The inner-loop of TDD is also commonly referred to as "Red-Green-Refactor"; red to represent a failing test, green to represent all passing tests, and refactoring only when all tests are passing.
+
+<%= partial 'partials/svg-tdd-red-green-refactor' %>
+
+### Red - write a failing test
```
Comment (huwd) [inline source/standards/test-driven-development.html.md.erb:18]:
```suggestion
### Red - write a failing test

```

File: source/standards/test-driven-development.html.md.erb [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+Turn exactly one item on the list into an actual, concrete, runnable test. This test should fail when you run it, and it should fail in the way you expect. If it fails unexpectedly, that could indicate a functionality gap or a misimplementation. For example, setting a field as immutable when you expect to mutate it on later interactions.
+
+### Green - make all tests pass
```
Comment (huwd) [inline source/standards/test-driven-development.html.md.erb:21]:
```suggestion
### Green - make all tests pass

```

File: source/standards/test-driven-development.html.md.erb [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+
+### Refactor - iterate on your design
+
+Improve the design and implementation of the code by removing duplication, utilising well known patterns and structures. You should let yourself be guided by the tests you've written; if a test is hard to write or you find yourself having to update many tests to support a new test without using automated refactoring tools, then take a step back and examine why.
+
+These steps are repeated until your set of behaviours is exhausted.
+
+## Why we recommend it
+
+The advantage of writing the tests before the implementation is that the behaviour of the code you will write is defined first, rather than implementing the change and retroactively applying tests.
```
Comment (huwd) [inline source/standards/test-driven-development.html.md.erb:32]: Do you think it's also worth mentioning TDD as an aid to thinking here?

Personally I find TDD forces me to confront a bunch of design questions with a degree of clarity that I otherwise miss. It's so easy to charge ahead on a problem with a fairly fuzzy idea in mind. TDD helps turn fuzzy dreams into firm plans.

**Review by huwd — APPROVED** [review]
Content looks great,

We might consider adding some links to some more pratical / worked examples?
I feel like folks can struggle translating the theory into that early practice of "simple test" -> "oh i didn't import from my lib file", "oh i didn't write a class", "oh i didn't initalise the class with the required parameter", "oh there's no method called `sum`" to get to that first green?

This page makes sense as a set of principles, but wonder if we can endorse a good worked example at the bottom that would help folks see the detail and build that muscle memory?

Can very much be added later.
Also a few nitpicks on Markdown formatting.

File: source/standards/test-driven-development.html.md.erb (hunk 6 above)
Comment (mrwilson) [reply source/standards/test-driven-development.html.md.erb:42]: I've put something about software design in the "refactor" section above.

File: source/partials/_svg-tdd-red-green-refactor.html.erb (hunk 3 above)
Comment (mrwilson) [reply source/partials/_svg-tdd-red-green-refactor.html.erb]: I've put the line break in but the requirement for .erb is a middleman thing - the build breaks if you remove it.

**Comment by mrwilson** [comment]
> I very much support adding TDD to The GDS Way! When I learned about TDD it was with a peer group who benefitted from a lot of time and support from our managers, which included regular, safe opportunities to practise our TDD skills with expert advice. I wonder if guidance is needed for those who are new to TDD and/or haven't had any training or coaching in it?

I completely agree - I've added some sections about "Getting started" and "Getting better" but I had a similar experience - pairing/mobbing and TDDing every day, which turbocharged my learning.

I'd like us to do more training / support for people to get the same experience I did (which I'm also happy to do).

**Comment by mrwilson** [comment]
> Content looks great,
> 
> We might consider adding some links to some more pratical / worked examples? I feel like folks can struggle translating the theory into that early practice of "simple test" -> "oh i didn't import from my lib file", "oh i didn't write a class", "oh i didn't initalise the class with the required parameter", "oh there's no method called `sum`" to get to that first green?
> 
> This page makes sense as a set of principles, but wonder if we can endorse a good worked example at the bottom that would help folks see the detail and build that muscle memory?
> 
> Can very much be added later. Also a few nitpicks on Markdown formatting.

Agree, I've linked two very good books (GOOS and By Example) but also Jason Gorman's page which has (a) his TDD book with many worked examples and (b) his 101 TDD Tips which is full of bitesize ... tips to improve your TDD.

**Comment by mboban2024** [comment]
I support this approach from a quality engineering and testing perspective. The content looks really good. To achieve good quality tests that drive development, it's crucial to create a testing mindset and culture. Without that TDD may fail to produce intended outcomes. Completely agree with principles and the value TDD brings, but it is important that teams don't misunderstand it as one-stop solution for all testing needs. Hence I think it’s worth also define what TDD is not, to clarify its boundaries or distinguish it from other types of testing in the software development lifecycle. A few suggestions:

- TDD is not a replacement for all test levels: TDD focuses on unit testing level and component integration. It doesn’t replace other levels of testing, such as functional integration testing, system testing, or feature acceptance testing. But good quality at unit testing reduces quality risk and extensive testing in higher levels
- TDD is not a replacement for exploratory testing: Exploratory testing is used to examine the application to find issues that automated tests may not capture. Exploratory testing uncovers unexpected behaviours outside the scope of TDD. It is a specialised task to prevent unknown - knowns (unaware of behaviour, but can exist in real usage) or unknown-unknowns (latent issues within our systems—problems we didn't even know could be problems or manifest when conditions line up just the right way) in the system.
- TDD is not a Catch-All for bugs: TDD helps prevent bugs in isolated units of code but does not catch all types of issues, especially those arising from system interactions or complex edge cases.
- TDD is more technically focused: Tests designed may miss view on business logic scenarios or user focused tests (else I think we need to experiment and evidence how module or class focused tests can have wider view on business and user needs).

**Comment by mrwilson** [comment]
Thanks @mboban2024! I will fold your suggestions into the content

**Review by two-first-names — DISMISSED** [review]

**Review by two-first-names — APPROVED** [review]
